### DISCUSSION TITLE
ES32-S3-12K errors when compiling

### DISCUSSION THREAD

**Discussion by wernizh** [tbnobody/OpenDTU#622] [Q&A]
Hallo

(english see below)

Das Programm für mein Board ESP32-S3-12K  will nicht kompilieren. 
Mit     (in platformio.ini)
         [platformio]
         default_envs = esp_s3_12k_kit
       
Gibt es mehrere Fehler bei:
    ResetReason.cpp:38:5: error: case value '4' not in enumerated type 'RESET_REASON'  (mit verschiedenen Case Nummern).  (siehe weiter unten)
 
Hab es mit Visual Studio Code und mit PlatformIO versucht. Bei beiden mit gleichem Problem.
Hab dann folgendes versucht:
             [platformio]
             default_envs = generic
             ---> Resultat:   Es kompiliert fehlerfrei

--- danach --- weiterer Versuch:
             [platformio]
             default_envs = olimex_esp32_poe
             ---> Resultat:   Es kompiliert fehlerfrei

--- danach --- weiterer Versuch:
             [platformio]
             default_envs = wemos_d1_mini32
             ---> Resultat:   Es kompiliert fehlerfrei

--- danach --- nochmals 
        [platformio]
        default_envs = esp_s3_12k_kit
        ---> Resultat:   damit geht es nie....  es kommt immer der Fehler unten

Die fehlerfrei kompilierten kann ich nicht hochladen da er reklamiert, dass diese für Chip ESP32 sind; ich aber einen ESP32-S3 habe.

Hat mir da jemand einen Tipp wie ich eine ESP32-S3 Version hinkriege? resp. wo das Problem mit dem Fehler ist?
Beispielprogramme für den ESP32-S3 kann ich kompilieren / laden / ausführen. Funktioniert problemlos.

Besten Dank

Werner

I get compile errors when compiling for 
        [platformio]
        default_envs = esp_s3_12k_kit
(see below)

- I can compile (with Visual Studio Code and with PlatformIO) when I use:
             [platformio]
             default_envs = generic

- it also works/compiles with:
             [platformio]
             default_envs = olimex_esp32_poe
            
   and 
             [platformio]
             default_envs = wemos_d1_mini32
             

But it does not compile with:
        [platformio]
        default_envs = esp_s3_12k_kit

- It does compile a while but then I get an error (with Visual Studio Code and with PlatformIO)
   It complains about errors in the file:  
            lib/ResetReason/src/ResetReason.cpp:38:5: error: case value '4' not in enumerated type 'RESET_REASON' .....
        (see below)

What are I'm doing wrong?

Thanks a lot, Werner

Processing esp_s3_12k_kit (board: esp32-s3-devkitc-1; framework: arduino; platform: espressif32@>=6.0.1)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
Firmware Revision: g97bc964
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32-s3-devkitc-1.html
PLATFORM: Espressif 32 (6.0.1) > Espressif ESP32-S3-DevKitC-1-N8 (8 MB QD, No PSRAM)
HARDWARE: ESP32S3 240MHz, 320KB RAM, 8MB Flash
DEBUG: Current (esp-builtin) On-board (esp-builtin) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 3.20006.221224 (2.0.6) 
 - tool-esptoolpy @ 1.40400.0 (4.4.0) 
 - tool-mkfatfs @ 2.0.1 
 - tool-mklittlefs @ 1.203.210628 (2.3) 
 - tool-mkspiffs @ 2.230.0 (2.30) 
 - tool-openocd-esp32 @ 2.1100.20220706 (11.0) 
 - toolchain-riscv32-esp @ 8.4.0+2021r2-patch5 
 - toolchain-xtensa-esp32s3 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 45 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ESP Async WebServer @ 1.2.3+sha.1dde9cf
|   |-- AsyncTCP @ 1.1.1
|   |-- ArduinoJson @ 6.20.1
|   |-- FS @ 2.0.0
|   |-- WiFi @ 2.0.0
|-- ArduinoJson @ 6.20.1
|-- espMqttClient @ 1.3.3+sha.6a3f3a8
|   |-- AsyncTCP @ 1.1.1
|   |-- WiFiClientSecure @ 2.0.0
|   |   |-- WiFi @ 2.0.0
|   |-- WiFi @ 2.0.0
|-- RF24 @ 1.4.5
|   |-- SPI @ 2.0.0
|-- U8g2 @ 2.34.13
|   |-- SPI @ 2.0.0
|   |-- Wire @ 2.0.0
|-- sunset @ 1.1.7
|-- LittleFS @ 2.0.0
|   |-- FS @ 2.0.0
|-- Hoymiles @ 0.0.1
|   |-- RF24 @ 1.4.5
|   |   |-- SPI @ 2.0.0
|   |-- Every
|   |-- SPI @ 2.0.0
|   |-- TimeoutHelper
|-- Ethernet @ 2.0.0
|   |-- WiFi @ 2.0.0
|-- Update @ 2.0.0
|-- ResetReason
|-- DNSServer @ 2.0.0
|   |-- WiFi @ 2.0.0
|-- WiFi @ 2.0.0
|-- MqttSubscribeParser
|   |-- espMqttClient @ 1.3.3+sha.6a3f3a8
|   |   |-- AsyncTCP @ 1.1.1
|   |   |-- WiFiClientSecure @ 2.0.0
|   |   |   |-- WiFi @ 2.0.0
|   |   |-- WiFi @ 2.0.0
|-- Ticker @ 2.0.0
Building in release mode
Converting .pio\build\esp_s3_12k_kit\index.html.gz.txt.o
Converting .pio\build\esp_s3_12k_kit\zones.json.gz.txt.o
Converting .pio\build\esp_s3_12k_kit\favicon.ico.txt.
.
.
.
.
Archiving .pio\build\esp_s3_12k_kit\lib664\libEthernet.a
Archiving .pio\build\esp_s3_12k_kit\libbde\libHoymiles.a
Compiling .pio\build\esp_s3_12k_kit\libd11\MqttSubscribeParser\MqttSubscribeParser.cpp.o
Compiling .pio\build\esp_s3_12k_kit\lib99b\Ticker\Ticker.cpp.o
lib/ResetReason/src/ResetReason.cpp: In member function 'String ResetReasonClass::get_reset_reason_verbose(uint8_t)':
lib/ResetReason/src/ResetReason.cpp:38:5: error: case value '4' not in enumerated type 'RESET_REASON' [-Werror=switch]
     case 4:
     ^~~~
lib/ResetReason/src/ResetReason.cpp:46:5: error: case value '6' not in enumerated type 'RESET_REASON' [-Werror=switch]
     case 6:
     ^~~~
lib/ResetReason/src/ResetReason.cpp:72:5: error: case value '14' not in enumerated type 'RESET_REASON' [-Werror=switch]
     case 14:
     ^~~~
lib/ResetReason/src/ResetReason.cpp: In member function 'String ResetReasonClass::get_reset_reason_short(uint8_t)':
lib/ResetReason/src/ResetReason.cpp:104:5: error: case value '4' not in enumerated type 'RESET_REASON' [-Werror=switch]
     case 4:
     ^~~~
lib/ResetReason/src/ResetReason.cpp:112:5: error: case value '6' not in enumerated type 'RESET_REASON' [-Werror=switch]
     case 6:
     ^~~~
lib/ResetReason/src/ResetReason.cpp:138:5: error: case value '14' not in enumerated type 'RESET_REASON' [-Werror=switch]
     case 14:
     ^~~~
cc1plus.exe: all warnings being treated as errors
Compiling .pio\build\esp_s3_12k_kit\FrameworkArduino\Esp.cpp.o
*** [.pio\build\esp_s3_12k_kit\lib1ec\ResetReason\ResetReason.cpp.o] Error 1

**Comment by tbnobody** [comment]
Will fix it tomorrow... If you can't wait, edit `lib/ResetReason/src/ResetReason.cpp` and replace all 
`#ifndef CONFIG_IDF_TARGET_ESP32C3`
by
`#if !defined(CONFIG_IDF_TARGET_ESP32C3) && !defined(CONFIG_IDF_TARGET_ESP32S3)`

But I don't own a ESP32-S3... I am not sure (even if the code compiles) if everything works without problems.

**Reply by wernizh** [reply]
Thanks a lot. 

Just before your answer I did try a more radical idea:    In  lib/ResetReason/src/ResetReason.cpp 
a) I deleted all the lines with #ifndef...     then 
b) I deleted all the lines with   case value '4'      case value '6'     case value '14'
(I'm aware that maybe I could get in trouble with other boards. So it was just a try.

With this it compiles and it works!!!!

Thanks a lot for this amazing project!

Have a nice day!

**Comment by einfacharthur** [comment]
Hi, I have sucessfully flashed openDTU on a ESP-WROOM-32 Board with the precompiled files:

bootloader.bin
boot_app0.bin
opendtu-generic.bin
partitions.bin

but I also have this ESP32-S3 Boards: https://www.wemos.cc/en/latest/s3/s3.html + https://de.aliexpress.com/item/1005004617322170.html
Lolin S3:
based ESP32-S3-WROOM-1
16MB Flash (Quad SPI)
8MB PSRAM (Octal SPI)

are this files compatible with the ESP32-S3 or C3 with the same settings as shown here:
![image](https://user-images.githubusercontent.com/12992757/221060773-7bb7c4e3-d29a-4cc3-a124-e4c05e43a619.png)
?

@MrReSc how you got it working?

Mit Putty kriege ich auch nur folgende Ausgabe:

invalid header: 0▒ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x10 (RTCWDT_RTC_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
invalid header: 0xffffffff


▒ESP-ROM:esp32s3-20210327f
Build:Mar 27 2021
rst:0x7 (TG0WDT_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40048836

invalid header▒ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x7 (TG0WDT_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40049b21

......

jemand ne Idee? :)

okay bin schon mal weitergekommen:

platformio.ini
default_envs = lolin_s3
statt
default_envs = generic

platformio_override.ini
[env:lolin_s3]
platform = espressif32
board = lolin_s3

; change microcontroller
board_build.mcu = esp32s3

; change MCU frequency
board_build.f_cpu = 240000000L

yeaaah flashed sucessfully..
![image](https://user-images.githubusercontent.com/12992757/221070242-e6ddb596-e62d-4885-b900-68ab7dc9afff.png)

but it still won't work..

**Reply by wernizh** [reply]
Hallo einfacharthur

Das ist bei mir mit meinem ESP32-S3-12K  board auch so.  Ich kann die precompiled Dateien problemlos hochladen. Bekomme dann aber auch "invalid header 0xffffffff).      Ich bin neu mit ESP32; kann mir aber vorstellen, dass die precompiled Dateien für ESP32 sind und nicht mit ESP32-S3 kompatibel sind???    auch habe ich irgendwo gelesen, dass das Partitionsschema bei den div. ESPs unterschiedlich ist. Möglicherweise muss man dort andere Bereiche angeben?

Wie auch immer. Bei mir hat es nun geklappt mit:
[platformio]
default_envs = esp_s3_12k_kit
(ohne weiter Parameter zu ändern oder zusätzlich hinzuzufügen. ) Allerdings musste ich dazu  in "lib/ResetReason/src/ResetReason.cpp" ein paar Zeilen rauslöschen (siehe mein anderer Kommentar).   

Schönen Tag und viel Erfolg!

Werner
